File: core_libraries/submodules/CNT_research_tools/Matlab/references/laplacian_reference.m
function [out_values,close_chs,laplacian_labels] = laplacian_reference(values,include,locs,radius,labels)

assert(~isempty(locs))

out_values = nan(size(values));
nchs = size(values,2);

D = pdist2(locs,locs); 
close = D < radius;
close(logical(eye(size(close)))) = 0;

close_chs = cell(nchs,1);

for i = 1:nchs
    out_values(:,i) = values(:,i) - nanmean(values(:,close(i,:)&include'),2);
    close_chs{i} = find(close(i,:)&include');
end

%laplacian_labels = strcat(labels,'-laplacian');
laplacian_labels = cell(length(labels),1);
for i = 1:length(laplacian_labels)
    ch_string = '';
    for j = 1:length(close_chs{i})
        ch_string = [ch_string,labels{close_chs{i}(j)}];
        if j < length(close_chs{i})
            ch_string = [ch_string,', '];
        end
    end
    laplacian_labels{i} = [labels{i},'-',ch_string];
end


end

File: core_libraries/subtrees/preprocessing_research_tool/python/tools/laplacian_reference.py
import numpy as np
from scipy.spatial.distance import cdist

def laplacian_reference(data:np.ndarray,locs,radius,labels):
    """_summary_

    Args:
        data (np.ndarray): _description_
        locs (_type_): _description_
        radius (_type_): _description_
        labels (_type_): _description_
    """
    print("Currently only available for test for HUP212_phaseII")
    out_values = np.nan*np.zeros(data.shape)
    nchs = data.shape[1]

# Calculate pairwise distances between X and Y
    D = cdist(locs, locs)

    close = D < radius
    close[np.eye(close.shape[0]).astype(bool)] = 0

    close_chs = []

    for i in range(nchs):
        out_values[:,i] = data[:,i] - np.nanmean(data[:,close[i,:]],1)
        close_chs.append(np.where(close[i,:])[0])

    # laplacian_labels = [i+'-laplacian' for i in labels]
    laplacian_labels = []
    for i in range(len(labels)):
        ch_string = '-'
        ch_list = []
        for j in range(len(close_chs[i])):
            ch_list.append(labels[close_chs[i][j]])
        ch_string = ch_string + ",".join(ch_list)
        laplacian_labels.append(labels[i] + ch_string)

    return out_values,close_chs,laplacian_labels